src/aes_cl12/aes_cl12.cc

```diff
@@ -1,373 +1,390 @@
-        {
-                uint8_t bytes[4];
-                uint32_t word;
-        } univar[60] __attribute__ ((aligned));
-
-        for (i = 0; i < Nk; i++)
-        {
-                univar[i].bytes[3] = pk[i*4];
-                univar[i].bytes[2] = pk[i*4+1];
-                univar[i].bytes[1] = pk[i*4+2];
-                univar[i].bytes[0] = pk[i*4+3];
-        }
+uint32_t AES::SubWord(uint32_t word) {
+  union {
+    uint32_t word;
+    uint8_t bytes[4];
+  } subWord  __attribute__((aligned));
 
-        for (i = Nk; i < Nb*(Nr+1); i++)
-        {
-                temp.word = univar[i-1].word;
-                if (i % Nk == 0)
-                {
-                    temp.word = (SubWord(RotateWord(temp.word)));
-                    temp.bytes[3] = temp.bytes[3] ^ (Rcon[i/Nk]);
-                }
-                else if (Nk > 6 && i % Nk == 4)
-                    temp.word = SubWord(temp.word);
-
-                if (i-4 % Nk == 0)
-                    temp.word = SubWord(temp.word);
-
-                univar[i].word = univar[i-Nk].word ^ temp.word;
-        }
+  subWord.word = word;
 
-        // Copy from the buffer into the variable
-        for (i = 0; i < 60; i++)
-                expanded_key[i] = univar[i].word;
+  subWord.bytes[3] = s[subWord.bytes[3]];
+  subWord.bytes[2] = s[subWord.bytes[2]];
+  subWord.bytes[1] = s[subWord.bytes[1]];
+  subWord.bytes[0] = s[subWord.bytes[0]];
 
+  return subWord.word;
 }
 
-void AES::InitKeys()
-{
-        // Read the private key in
-        for (int i = 0; i < 32; i++)
-                int res = fscanf(keyfile, "%02x", (int *)&key[i]);
-                //res is simply a variable to supress the "ignored"
-                //results warning
-        // Expand key
-        KeyExpansion(key);
-
+void AES::KeyExpansion(uint8_t *pk) {
+  int i = 0;
+  // Temp union will hold the word that is being processed
+  union {
+    uint8_t bytes[4];
+    uint32_t word;
+  } temp __attribute__((aligned));
+
+  // Univar is the buffer that will hold the expanded key, it
+  // is loaded in 8-bit parts which is why the union is necessary
+  union {
+    uint8_t bytes[4];
+    uint32_t word;
+  } univar[60] __attribute__((aligned));
+
+  for (i = 0; i < Nk; i++) {
+    univar[i].bytes[3] = pk[i*4];
+    univar[i].bytes[2] = pk[i*4+1];
+    univar[i].bytes[1] = pk[i*4+2];
+    univar[i].bytes[0] = pk[i*4+3];
+  }
+
+  for (i = Nk; i < Nb*(Nr+1); i++) {
+    temp.word = univar[i-1].word;
+    if (i % Nk == 0) {
+      temp.word = (SubWord(RotateWord(temp.word)));
+      temp.bytes[3] = temp.bytes[3] ^ (Rcon[i/Nk]);
+    } else if (Nk > 6 && i % Nk == 4) {
+      temp.word = SubWord(temp.word);
+    }
+    if (i-4 % Nk == 0)
+      temp.word = SubWord(temp.word);
+
+    univar[i].word = univar[i-Nk].word ^ temp.word;
+  }
+
+  // Copy from the buffer into the variable
+  for (i = 0; i < 60; i++)
+    expanded_key[i] = univar[i].word;
 }
 
-void AES::Run(int argc, char const *argv[])
-{
-        cl_int err;
```

Comment:
Avoid putting everything in the Run function. You have Initialize function for initializing all the runtime objects, buffers. Use may only care about the algorithm execution time, but not caring about initializing time. Keep them separate.